File: c++/leetcode/1381. Design a Stack With Increment Operation.cpp
class CustomStack
{
  int size;
  int capacity;
  struct Node
  {
    int data;
    Node *next;
    Node(int data) : data{data} {};
  } * head;

public:
  CustomStack(int maxSize)
  {
    capacity = maxSize;
    head = nullptr;
    size = 0;
  }

  void push(int x)
  {
    Node *latest = new Node{x};
    if (size != capacity)
    {
      latest->next = head;
      head = latest;
      size++;
    }
  }

  int pop()
  {
    if (!size)
      return -1;
    Node *to_delete = head;
    int val = to_delete->data;
    head = head->next;
    delete to_delete;
    size--;
    return val;
  }

  void increment(int k, int val)
  {
    vector<int> temp;
    Node *nav = head;
    if (size < k)
    {
      // do it for everything in stack
      while (nav)
      {
        nav->data += val;
        nav = nav->next;
      }
    }
    else
    {

      for (int i = 1; i <= size - k; i++)
      {
        nav = nav->next;
      }

      while (nav)
      {
        nav->data += val;
        nav = nav->next;
      }
    }
  }
};

/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack* obj = new CustomStack(maxSize);
 * obj->push(x);
 * int param_2 = obj->pop();
 * obj->increment(k,val);
 */

File: javascript/leetcode/1381. Design a Stack With Increment Operation.js
/**
 * @param {number} maxSize
 */
var CustomStack = function(maxSize) {
  this.capacity = maxSize;
  this.size = 0;
  this.node = [];
};

/**
 * @param {number} x
 * @return {void}
 */
CustomStack.prototype.push = function(x) {
  if (this.size != this.capacity) {
    this.node.push(x);
    this.size++;
  }
};

/**
 * @return {number}
 */
CustomStack.prototype.pop = function() {
  if (!this.size) return -1;
  this.size--;
  return this.node.pop();
};

/**
 * @param {number} k
 * @param {number} val
 * @return {void}
 */
CustomStack.prototype.increment = function(k, val) {
  let end = this.size;
  if (!(this.size < k)) end = k;
  for (let i = 0; i < end; i++) this.node[i] += val;
};

/**
 * Your CustomStack object will be instantiated and called as such:
 * var obj = new CustomStack(maxSize)
 * obj.push(x)
 * var param_2 = obj.pop()
 * obj.increment(k,val)
 */
